### PR TITLE
Add verification on mergeSpans

### PR DESCRIPTION
**Pull request by ahrtr**
It might be helpful to detect issue such as https://github.com/etcd-io/bbolt/issues/705

cc @fuweid @tjungblu

### CODE REVIEW

File: freelist.go [hunk 1]
```diff
@@ -389,5 +389,29 @@ func (f *freelist) reindex() {
...
+			if _, ok := idsIdx[id]; ok {
+				panic(fmt.Sprintf("detected duplicated free page ID: %d in existing f.ids: %v", id, f.ids))
+			}
+			idsIdx[id] = struct{}{}
+		}
+
+		prev := common.Pgid(0)
+		for _, id := range ids {
+			// The ids shouldn't have duplicated free ID.
+			if prev == id {
```
Comment (tjungblu) [inline freelist.go:405]: maybe leave a little comment here that the meta page 0 can never be in the free list, otherwise you'll potentially panic here

File: freelist_hmap.go [hunk 2]
```diff
@@ -108,6 +110,33 @@ func (f *freelist) hashmapGetFreePageIDs() []common.Pgid {
 
 // hashmapMergeSpans try to merge list of pages(represented by pgids) with existing spans
 func (f *freelist) hashmapMergeSpans(ids common.Pgids) {
+	common.Verify(func() {
+		ids1 := f.idsFromFreemaps()
+		ids2 := f.idsFromForwardMap()
+		ids3 := f.idsFromBackwardMap()
```
Comment (tjungblu) [inline freelist_hmap.go:116]: could you give those variables a proper name please? it's a bit hard to follow the logic below without the backward/forward in mind

File: freelist.go (hunk 1 above)
Comment (ahrtr) [reply freelist.go:406]: Updated the comment.

Comment (tjungblu) [reply freelist.go:406]: thx!

File: freelist_hmap.go (hunk 2 above)
Comment (ahrtr) [reply freelist_hmap.go]: Updated. I was thinking the short names `ids1/ids2/ids3` are better.

**Comment by tjungblu** [comment]
/lgtm

thanks for adding those, let's see if we can find some triggers of this

**Comment by tjungblu** [comment]
this is highly off topic here, but I wonder if we can also implement a much simpler freelist with plain bitsets. We would have a much more compact in-memory representation and would get unions/intersections for free.

Had some good experience with github.com/bits-and-blooms/bitset recently while creating that defrag/page usage video.

**Comment by ahrtr** [comment]
> this is highly off topic here, but I wonder if we can also implement a much simpler freelist with plain bitsets. We would have a much more compact in-memory representation and would get unions/intersections for free.
> 
> Had some good experience with github.com/bits-and-blooms/bitset recently while creating that defrag/page usage video.

I am not satisfied with the bbolt freelist management for a long time. I have been thinking how to refactor it and improve the situation.

I am open to any idea on this and happy to discuss it further. Will take a look at github.com/bits-and-blooms/bitset later.

**Comment by ahrtr** [comment]
> I am not satisfied with the bbolt freelist management for a long time

There isn't a clear interface and hard to understand. I am thinking to 
- define a clear interface & clarify the behavior.
- try to decouple it with the bbolt;
- add more dedicated test for the freelist management

**Comment by tjungblu** [comment]
yep, that would also be my first course of action. Pull the two implementations under a common interface, then have proper testing for those.

**Comment by ahrtr** [comment]
> github.com/bits-and-blooms/bitset

I just had a quick look at the project. It seems not quite related to our usage o freelist management. It manages a mapping between non-negative integers and boolean values; while bbolt freelist management just needs to remember a list of free page IDs.

If you still think it might be helpful or related, please feel free to raise a separate ticket to discuss. thx

**Comment by tjungblu** [comment]
> If you still think it might be helpful or related, please feel free to raise a separate ticket to discuss. thx

will do. You can use a bitset to map from PageId->Free/Used. Since you're packing 64 bools into one int64, you have a much lower memory/file usage. Finding continuous blocks is also cheaper, since you can just bitshift along the int64.
